Improved tearDown methods for SmartClass/SmartVariables tests.

svtkachenko opened the pull request:
Sometitimes it was not possible to remove puppet class because it was associated with hosts/hostgroups, however these testcases depend on puppet class state. Closes #3863

Changes:
* Improved tearDownClass methods
* Updated CLI/SmartVariables tests according to changes in 6.3
* Updated CLI/SmartClassParameters tests according to changes in 6.3

TearDown in CLI uses nailgun and API calls for several reasons:
1. Looks like it is not possible to delete specific puppet class from host/hostgroup using hammer. Asking on #theforeman I've found that it is possible only to replace old list of classes wih new list which doesn't include class that should be removed using `host/hostgroup update`. But in case of empty list this is does not work.
2. Due to opend BZ `hostgroup update` does not work in hammer.

Test results for API/CLI
```python
% nosetests tests/foreman/api/test_variables.py
.....S..S..SSSSSS........S............SSSSS
----------------------------------------------------------------------
Ran 43 tests in 147.220s
OK (SKIP=14)

% nosetests tests/foreman/api/test_classparameters.py
SSSS.............SSSSSS.....S..S..............
----------------------------------------------------------------------
Ran 46 tests in 145.698s
OK (SKIP=12)

% nosetests tests/foreman/cli/test_classparameters.py
.SSSS..........SSSSSS..........S...........
----------------------------------------------------------------------
Ran 43 tests in 671.070s
OK (SKIP=11)

% nosetests tests/foreman/cli/test_variables.py 
..SSSS..SSS..S......SSSSS.....S...........S......
----------------------------------------------------------------------
Ran 49 tests in 972.346s
OK (SKIP=15)
```
As tearDownClass should not affect tests too much and as some UI tests failing because of 'blue' interface of Snap3, only one test result as example for UI.
```python
% pytest tests/foreman/ui/test_variables.py -k 'test_positive_create_with_host'
===================================================================== test session starts =====================================================================
platform linux2 -- Python 2.7.12, pytest-3.0.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/qui/code/robottelo, inifile: 
plugins: cov-2.3.1, xdist-1.15.0
collected 44 items 

tests/foreman/ui/test_variables.py .

===================================================================== 43 tests deselected =====================================================================
========================================================== 1 passed, 43 deselected in 113.90 seconds ==========================================================
```

svtkachenko commented on tests/foreman/cli/test_classparameters.py:
```diff
@@ -1001,7 +1030,6 @@ def test_positive_create_matcher_puppet_default_value(self):
         SmartClassParameter.add_override_value({
             'smart-class-parameter-id': sc_param_id,
             'match': 'domain=test.com',
-            'value': value,
```
Change in hammer:
```
Could not create the override value:
  Error: Cannot use --value when --use-puppet-default is true
    See: 'hammer sc-param add-override-value --help'
```

sghai commented:
ACK.. Looks good.

@jyejare : Could you please take a look ? and do you know about these changes @svtkachenko  pointed ?

jyejare commented:
Looks Good to me as well.
Also thanks @sghai for pointing me to this PR, I was unaware of these changes !

oshtaier commented:
It seems that you are adding completely identical pieces of code in different modules. Could you please create a helper or something?

svtkachenko commented:
Updated: added helper that removes puppet class entity and puppet module using nailgun, though where it is possible hammer is used to be unifom with setup method.

API:
```python
% pytest -v tests/foreman/api/test_classparameters.py -k 'test_positive_list_parameters_by_host_id or test_positive_list_parameters_by_hostgroup_id'                       
===================================================================== test session starts =====================================================================
tests/foreman/api/test_classparameters.py::SmartClassParametersTestCase::test_positive_list_parameters_by_host_id <- robottelo/decorators/__init__.py PASSED
tests/foreman/api/test_classparameters.py::SmartClassParametersTestCase::test_positive_list_parameters_by_hostgroup_id <- robottelo/decorators/__init__.py PASSED

===================================================================== 45 tests deselected =====================================================================
========================================================== 2 passed, 45 deselected in 41.40 seconds ===========================================================
```
```python
% pytest -v tests/foreman/api/test_variables.py -k 'test_positive_list_variables_by_host_id or test_positive_list_variables_by_hostgroup_id'                                                                                           
===================================================================== test session starts =====================================================================
tests/foreman/api/test_variables.py::SmartVariablesTestCase::test_positive_list_variables_by_host_id <- robottelo/decorators/__init__.py PASSED
tests/foreman/api/test_variables.py::SmartVariablesTestCase::test_positive_list_variables_by_hostgroup_id <- robottelo/decorators/__init__.py PASSED

===================================================================== 41 tests deselected =====================================================================
========================================================== 2 passed, 41 deselected in 42.01 seconds ===========================================================
```
CLI:
```python
% pytest -v tests/foreman/cli/test_classparameters.py -k 'test_positive_list_parameters_by_host_id or test_positive_list_parameters_by_hostgroup_id'                                        
===================================================================== test session starts =====================================================================
tests/foreman/cli/test_classparameters.py::SmartClassParametersTestCase::test_positive_list_parameters_by_host_id <- robottelo/decorators/__init__.py PASSED
tests/foreman/cli/test_classparameters.py::SmartClassParametersTestCase::test_positive_list_parameters_by_hostgroup_id <- robottelo/decorators/__init__.py PASSED

===================================================================== 41 tests deselected =====================================================================
========================================================== 2 passed, 41 deselected in 111.16 seconds ==========================================================
```
```python
% pytest -v tests/foreman/cli/test_variables.py -k 'test_positive_list_variables_by_host_id or test_positive_list_variables_by_hostgroup_id'                                                
===================================================================== test session starts =====================================================================
tests/foreman/cli/test_variables.py::SmartVariablesTestCase::test_positive_list_variables_by_host_id <- robottelo/decorators/__init__.py PASSED
tests/foreman/cli/test_variables.py::SmartVariablesTestCase::test_positive_list_variables_by_hostgroup_id <- robottelo/decorators/__init__.py PASSED

===================================================================== 47 tests deselected =====================================================================
========================================================== 2 passed, 47 deselected in 107.21 seconds ==========================================================
```
UI:
```python
% pytest tests/foreman/ui/test_variables.py -k 'test_positive_create_with_host'                 [git][robottelo/.][f_tdU]
===================================================================== test session starts =====================================================================
tests/foreman/ui/test_variables.py .

===================================================================== 43 tests deselected =====================================================================
========================================================== 1 passed, 43 deselected in 123.21 seconds ==========================================================
```

oshtaier commented:
ACK
